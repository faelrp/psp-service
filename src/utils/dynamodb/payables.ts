import DynamoDB from './connector';
import config from '../../config';

const documentClient = DynamoDB.getDocumentClient();

export enum PayablesType {
  PAID = 'paid',
  WAITING_FUNDS = 'waiting_funds',
}

export interface Payable {
  cardNumber: string;
  tnxHash: string;
  status: PayablesType;
  amount: number;
  paymentDate: string;
  transactionAmount: number;
}

interface Add {
  payable: Payable;
}

const add = async ({ payable }: Add) => {
  const {
    aws: {
      dynamodb: { payablesTable },
    },
  } = config;

  const ts = new Date().getTime();
  const params = {
    TableName: payablesTable,
    Item: {
      ...payable,
      ts,
    },
    ConditionExpression: 'attribute_not_exists(#tnxHash)',
    ExpressionAttributeNames: { '#tnxHash': 'tnxHash' },
  };

  await documentClient.put(params).promise();
};

export default {
  add,
};
