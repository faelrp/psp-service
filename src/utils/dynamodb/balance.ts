import DynamoDB from './connector';
import config from '../../config';

import { PayablesType } from './payables';

const documentClient = DynamoDB.getDocumentClient();

export interface BalanceEvent {
  status: PayablesType;
  amount: number;
}

interface InitParams {
  status: PayablesType;
}

const {
  aws: {
    dynamodb: { payablesBalanceTable },
  },
} = config;

const init = async (balance: InitParams) => {
  const params = {
    TableName: payablesBalanceTable,
    Item: {
      ...balance,
      amount: 0,
    },
    ConditionExpression: 'attribute_not_exists(#status)',
    ExpressionAttributeNames: { '#status': 'status' },
  };

  try {
    await documentClient.put(params).promise();
  } catch (error) {
    // ConditionalCheckFailedException means that the balance was
    // already initiated, so its good to go.
    if (error.code !== 'ConditionalCheckFailedException') {
      throw error;
    }
  }
};

const update = async ({ status, amount }: BalanceEvent) => {
  await init({ status });

  const params = {
    TableName: payablesBalanceTable,
    Key: {
      status,
    },
    UpdateExpression: 'set amount = amount + :value',
    ExpressionAttributeValues: { ':value': Number(amount) },
  };

  await documentClient.update(params).promise();
};

const get = async ({ status }) => {
  const params = {
    TableName: payablesBalanceTable,
    Key: {
      status,
    },
  };

  const { Item } = await documentClient.get(params).promise();

  return Item || {};
};

export default {
  init,
  update,
  get,
};
