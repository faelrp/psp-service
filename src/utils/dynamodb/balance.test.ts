import balance from './balance';
import DynamoDB from './connector';
import { PayablesType } from './payables';

jest.mock('./connector');

// @ts-ignore
const dynamoPromise = DynamoDB.getPromise();
const { put, update, get } = DynamoDB.getDocumentClient();

describe('DynamoDb - Balance', () => {
  beforeAll(() => dynamoPromise.mockResolvedValue());
  afterEach(() => jest.clearAllMocks());

  describe('init', () => {
    test('Happy path should work', async () => {
      await balance.init({ status: PayablesType.PAID });

      expect(put).toHaveBeenCalledTimes(1);
      // @ts-ignore
      expect(put.mock.calls).toMatchSnapshot();
    });
  });
  describe('update', () => {
    test('Happy path should work', async () => {
      await balance.update({ status: PayablesType.PAID, amount: 10 });

      expect(update).toHaveBeenCalledTimes(1);
      // @ts-ignore
      expect(update.mock.calls).toMatchSnapshot();
    });
  });
  describe('get', () => {
    test('Happy path should work', async () => {
      dynamoPromise.mockResolvedValue({ Item: { unit: 'test' } });
      const balanceResult = balance.get({ status: PayablesType.PAID });

      expect(balanceResult).resolves.toMatchSnapshot();

      expect(get).toHaveBeenCalledTimes(1);
      // @ts-ignore
      expect(get.mock.calls).toMatchSnapshot();
    });
  });
});
