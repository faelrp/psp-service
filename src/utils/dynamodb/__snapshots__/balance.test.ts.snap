// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamoDb - Balance get Happy path should work 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "status": "paid",
      },
      "TableName": "payables_balance_dev",
    },
  ],
]
`;

exports[`DynamoDb - Balance get Happy path should work 2`] = `
Object {
  "unit": "test",
}
`;

exports[`DynamoDb - Balance init Happy path should work 1`] = `
Array [
  Array [
    Object {
      "ConditionExpression": "attribute_not_exists(#status)",
      "ExpressionAttributeNames": Object {
        "#status": "status",
      },
      "Item": Object {
        "amount": 0,
        "status": "paid",
      },
      "TableName": "payables_balance_dev",
    },
  ],
]
`;

exports[`DynamoDb - Balance update Happy path should work 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeValues": Object {
        ":value": 10,
      },
      "Key": Object {
        "status": "paid",
      },
      "TableName": "payables_balance_dev",
      "UpdateExpression": "set amount = amount + :value",
    },
  ],
]
`;
